File: problems/java/ParkingLotSystem/src/strategy/parking/NearestFirstStrategy.java
package strategy.parking;

import model.ParkingFloor;
import model.ParkingSpot;
import model.Vehicle;

import java.util.List;

public class NearestFirstStrategy implements ParkingStrategy {

    @Override
    public ParkingSpot findSpot(List<ParkingFloor> floors, Vehicle vehicle) {
        for (ParkingFloor parkingFloor : floors) {
            for (ParkingSpot parkingSpot : parkingFloor.getSpots()) {
                if (parkingSpot.canFitVehicle(vehicle)) {
                    return parkingSpot;
                }
            }
        }
        return null;
    }
}

File: problems/python/ParkingLotSystem/src/strategy/parking/nearest_first_strategy.py
"""Nearest-first parking strategy."""
from __future__ import annotations
from typing import Optional
from model.parking_floor import ParkingFloor
from model.parking_spot import ParkingSpot
from model.vehicle import Vehicle
from strategy.parking.parking_strategy import ParkingStrategy


class NearestFirstStrategy(ParkingStrategy):
    def find_spot(
        self, floors: list[ParkingFloor], vehicle: Vehicle
    ) -> Optional[ParkingSpot]:
        for floor in floors:
            for spot in floor.get_spots():
                if spot.can_fit_vehicle(vehicle):
                    return spot
        return None
